[JSC] Implement ArrayBuffer#transfer

Constellation opened the pull request:
#### fe0aa79252484e5da584cf525ccc930e7698deef
<pre>
[JSC] Implement ArrayBuffer#transfer
<a href="https://bugs.webkit.org/show_bug.cgi?id=263518">https://bugs.webkit.org/show_bug.cgi?id=263518</a>
rdar://117337535

Reviewed by Mark Lam and Ross Kirsling.

This patch implements ArrayBuffer#transfer proposal[1], which is stage-3 and implemented in the other engines too.
It implements two functions and one getter.

1. ArrayBuffer#transfer. It transfers internal ArrayBuffer to a new ArrayBuffer and detach the original buffer.
   This only works with normal ArrayBuffer and it throws an error against SharedArrayBuffer. This preserves resizability
   of the original ArrayBuffer, it means that if we transfer resizable ArrayBuffer, we get resizable transferred ArrayBuffer.
2. ArrayBuffer#transferToFixedLength is similar to ArrayBuffer#transfer. But this always creates non-resizable ArrayBuffer.
   It means, internally, we create a new ArrayBuffer if resizability is different and copy the underlying content.

In both cases, if ArrayBuffer is not resizable and result size is the same, then it just changes underlying buffer&apos;s ownership and
does not allocate and copy contents. If it is ArrayBuffer#transfer and the input ArrayBuffer is resizable, then we also just changes
the ownership and resize the resulted ArrayBuffer&apos;s size.

3. ArrayBuffer#detach getter returns boolean whether ArrayBuffer is detached or not.

[1]: <a href="https://tc39.es/proposal-arraybuffer-transfer/">https://tc39.es/proposal-arraybuffer-transfer/</a>

* JSTests/test262/config.yaml:
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/JSArrayBufferPrototype.cpp:
(JSC::arrayBufferCopyAndDetach):
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSArrayBufferPrototype::finishCreation):

Canonical link: <a href="https://commits.webkit.org/269674@main">https://commits.webkit.org/269674@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/86da10ff8fca597a0d98e9e4d534d25a1b460dab

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23263 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1376 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24390 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/25569 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/21491 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/2895 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23824 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/25569 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23503 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/936 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/20154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26035 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/750 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/21041 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/27199 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/20245 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/21209 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/21299 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/25074 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/22625 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/748 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/18504 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/30005 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/698 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6201 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/1163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/29957 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2953 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/999 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/6094 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/runtime/JSArrayBufferPrototype.cpp:
```diff
@@ -293,10 +295,75 @@ JSC_DEFINE_HOST_FUNCTION(arrayBufferProtoFuncResize, (JSGlobalObject* globalObje
+}
+
+// https://tc39.es/proposal-arraybuffer-transfer/#sec-arraybuffer.prototype.transfer
+JSC_DEFINE_HOST_FUNCTION(arrayBufferProtoFuncTransfer, (JSGlobalObject* globalObject, CallFrame* callFrame))
```
Since `arrayBufferProtoFuncTransfer` and `arrayBufferProtoFuncTransferToFixedLength` are literally identical except that they pass a different `CopyAndDetachMode` value to `arrayBufferCopyAndDetach` at the end, can you please create a common static function that both call with the `CopyAndDetachMode` as an argument?  Let's not copy paste code like this unnecessarily.  

In fact, it looks like the spec itself is doing this i.e. sharing code except for just the `CopyAndDetachMode` argument.  So, it looks like you can move most of this common functionality into `arrayBufferCopyAndDetach`.

Constellation replied:
I deliberately did this since I don't want to put these kind of argument handlings into `arrayBufferCopyAndDetach` and I would like to eventually move this to `ArrayBuffer::copyAndDetach`, which will take these arguments. I would like to keep it as is intentionally. I think the current form is better for the future ArrayBuffer::copyAndDetach.

MenloDorian replied:
In that case, can't you have a `arrayBufferProtoFuncTransferImpl` that calls `arrayBufferCopyAndDetach` instead?  Have `arrayBufferProtoFuncTransfer` and `arrayBufferProtoFuncTransferToFixedLength` call `arrayBufferProtoFuncTransferImpl` with a different `CopyAndDetachMode`.

Constellation replied:
Hmm, TBH, I don't agree that this makes code better..., or rather, it adds unnecessary complicated abstraction. But if you strongly think, I can just change.

MenloDorian reviewed (approved):
r=me with code sharing fix.

rkirsling commented on Source/JavaScriptCore/runtime/JSArrayBufferPrototype.cpp:
```diff
@@ -293,10 +295,75 @@ JSC_DEFINE_HOST_FUNCTION(arrayBufferProtoFuncResize, (JSGlobalObject* globalObje
+        // We should just transfer!
+        ArrayBufferContents contents;
+        if (UNLIKELY(!arrayBuffer->impl()->transferTo(vm, contents))) {
+            throwVMRangeError(globalObject, scope, makeString("transfer failed"_s));
```
I don't think you need `makeString` here? (And likewise for the other "transfer failed" cases.)

Also, the error messages would probably be nicer to read if we say "ArrayBuffer transfer" and "ArrayBuffer resize" instead of just "transfer" and "resize".

Constellation replied:
Fixed.

rkirsling commented on Source/JavaScriptCore/runtime/JSArrayBufferPrototype.cpp:
```diff
@@ -373,6 +455,19 @@ JSC_DEFINE_HOST_FUNCTION(arrayBufferProtoGetterFuncMaxByteLength, (JSGlobalObjec
+
+    auto* thisObject = jsDynamicCast<JSArrayBuffer*>(callFrame->thisValue());
+    if (!thisObject || (ArrayBufferSharingMode::Shared == thisObject->impl()->sharingMode()))
+        return throwVMTypeError(globalObject, scope, makeString("Receiver must be ArrayBuffer"_s));
```
Ditto about not needing `makeString`.

Constellation replied:
Fixed.

rkirsling reviewed (approved):
r=me too

Constellation committed:
[JSC] Implement ArrayBuffer#transfer
https://bugs.webkit.org/show_bug.cgi?id=263518
rdar://117337535

Reviewed by Mark Lam and Ross Kirsling.

This patch implements ArrayBuffer#transfer proposal[1], which is stage-3 and implemented in the other engines too.
It implements two functions and one getter.

1. ArrayBuffer#transfer. It transfers internal ArrayBuffer to a new ArrayBuffer and detach the original buffer.
   This only works with normal ArrayBuffer and it throws an error against SharedArrayBuffer. This preserves resizability
   of the original ArrayBuffer, it means that if we transfer resizable ArrayBuffer, we get resizable transferred ArrayBuffer.
2. ArrayBuffer#transferToFixedLength is similar to ArrayBuffer#transfer. But this always creates non-resizable ArrayBuffer.
   It means, internally, we create a new ArrayBuffer if resizability is different and copy the underlying content.

In both cases, if ArrayBuffer is not resizable and result size is the same, then it just changes underlying buffer's ownership and
does not allocate and copy contents. If it is ArrayBuffer#transfer and the input ArrayBuffer is resizable, then we also just changes
the ownership and resize the resulted ArrayBuffer's size.

3. ArrayBuffer#detach getter returns boolean whether ArrayBuffer is detached or not.

[1]: https://tc39.es/proposal-arraybuffer-transfer/

* JSTests/test262/config.yaml:
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/JSArrayBufferPrototype.cpp:
(JSC::arrayBufferCopyAndDetach):
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSArrayBufferPrototype::finishCreation):

Canonical link: https://commits.webkit.org/269674@main
